### PR TITLE
Fill in some missing parts in the default AST and HIR visitors

### PR DESCRIPTION
**Pull request by petrochenkov**
Some minor parts of AST and HIR were not visited by the `visit::walk_xxx` methods - some identifiers, lifetimes, loop labels, attributes of exported macros - but nothing as serious as in, for example, https://github.com/rust-lang/rust/pull/28364.
+ Added a convenience macro for visiting lists (including Options)
+ Removed some pre-Deref-coersions `&**` noise from visitors

r? @nrc

### CODE REVIEW

File: src/librustc_front/visit.rs [hunk 1]
```diff
@@ -309,91 +302,63 @@ pub fn walk_variant<'v, V: Visitor<'v>>(visitor: &mut V,
...
-                visitor.visit_ty(&*argument.ty)
-            }
-            walk_fn_ret_ty(visitor, &function_declaration.decl.output);
-            walk_lifetime_decls_helper(visitor, &function_declaration.lifetimes);
+            walk_fn_decl(visitor, &function_declaration.decl);
+            walk_list!(visitor, visit_lifetime_def, &function_declaration.lifetimes);
         }
         TyPath(ref maybe_qself, ref path) => {
-            if let Some(ref qself) = *maybe_qself {
+            for qself in maybe_qself {
```
Comment (nrc) [inline src/librustc_front/visit.rs]: Not sure I like this change - `if let` seems more appropriate than `for` here - why the change?

File: src/librustc_front/visit.rs [hunk 2]
```diff
@@ -447,17 +401,15 @@ pub fn walk_path_parameters<'v, V: Visitor<'v>>(visitor: &mut V,
...
 pub fn walk_pat<'v, V: Visitor<'v>>(visitor: &mut V, pattern: &'v Pat) {
     match pattern.node {
-        PatEnum(ref path, ref children) => {
+        PatEnum(ref path, ref opt_children) => {
             visitor.visit_path(path, pattern.id);
-            if let Some(ref children) = *children {
-                for child in children {
-                    visitor.visit_pat(&*child)
-                }
+            for children in opt_children {
```
Comment (nrc) [inline src/librustc_front/visit.rs]: same

**Comment by nrc** [comment]
Nice! `walk_list!` in particular is great.

r=me without the `if let` -> `for` changes.

File: src/librustc_front/visit.rs (hunk 1 above)
Comment (petrochenkov) [inline src/librustc_front/visit.rs]: It's just.. shorter, and doesn't have `()=*`s

**Comment by petrochenkov** [comment]
@nrc 
Updated.
